File: app/src/main/java/dev/voleum/ordermolder/adapters/CatalogListRecyclerViewAdapter.java
package dev.voleum.ordermolder.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.voleum.ordermolder.R;
import dev.voleum.ordermolder.databinding.CatalogHolderBinding;
import dev.voleum.ordermolder.models.Catalog;
import dev.voleum.ordermolder.viewmodels.ObjListItemViewModel;

public class CatalogListRecyclerViewAdapter<T extends Catalog> extends AbstractRecyclerViewAdapter<T> {

    private OnEntryClickListener onEntryClickListener;

    public CatalogListRecyclerViewAdapter(List<T> list) {
        super(list);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CatalogHolderBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                R.layout.catalog_holder,
                parent,
                false);
        return new CatalogViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ((CatalogViewHolder) holder).binding.setRow(new ObjListItemViewModel<>((list.get(position))));
    }

    public void setOnEntryClickListener(OnEntryClickListener onEntryClickListener) {
        this.onEntryClickListener = onEntryClickListener;
    }

    public class CatalogViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        CatalogHolderBinding binding;
        CatalogViewHolder(CatalogHolderBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            this.binding.getRoot().setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (onEntryClickListener != null) {
                onEntryClickListener.onEntryClick(v, getLayoutPosition());
            }
        }
    }

    public interface OnEntryClickListener {
        void onEntryClick(View v, int position);
    }
}

File: app/src/main/java/dev/voleum/ordermolder/adapters/CatalogListRecyclerViewAdapter.kt
package dev.voleum.ordermolder.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import dev.voleum.ordermolder.R
import dev.voleum.ordermolder.databinding.CatalogHolderBinding
import dev.voleum.ordermolder.models.Catalog
import dev.voleum.ordermolder.viewmodels.ObjListItemViewModel

class CatalogListRecyclerViewAdapter<T : Catalog>(list: List<T>) : AbstractRecyclerViewAdapter<T>(list) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding: CatalogHolderBinding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.context),
                        R.layout.catalog_holder,
                        parent,
                        false)
        return CatalogViewHolder(binding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        ((holder as CatalogListRecyclerViewAdapter<*>.CatalogViewHolder)
                .binding as CatalogHolderBinding)
                .row = ObjListItemViewModel(list[position])
    }

    inner class CatalogViewHolder(binding: CatalogHolderBinding) : AbstractViewHolder(binding) {

        init {
            binding.root.setOnClickListener { v -> onEntryClickListener.onEntryClick(v, layoutPosition) }
        }
    }
}